File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
input = sys.stdin.readline
N,C = map(int,input().split())
li=sorted([int(input()) for _ in range(N)])
start,end=1,li[-1]-li[0]
answer = 0

if C==2:
    print(li[-1]-li[0])
else:
    while(start<end):
        mid = (start+end)//2
        count = 1
        ts = li[0]
        
        for i in range(N):
            if li[i]-ts>=mid:
                count+=1
                ts=li[i]
        if count >= C:
            answer = mid
            start = mid + 1
        else:
            end = mid
    print(answer)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java

import java.io.*;
import java.util.*;
public class Main {
    static int N,C;
    static int[] map;
    public static void main(String[] args) throws IOException {
        input();
        Arrays.sort(map);

        int left=0;
        int right = map[N - 1] - map[0] + 1;
        while (left < right){
            int mid = left + (right-left)/2;
            if (can(mid) < C){
                right = mid;
            } else {
                left = mid +1;
            }
        }
        System.out.println(left-1);

    }
    static int can(int dist){
        int cnt = 1;
        int lastLocate = map[0];
        for (int i=1; i<N;i++){
            int currLocate = map[i];

            if (currLocate - lastLocate >= dist){
                cnt ++;
                lastLocate = currLocate;
            }
        }
        return cnt;
    }
    static void input() throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] arr = br.readLine().split(" ");
        N = Integer.parseInt(arr[0]);
        C = Integer.parseInt(arr[1]);
        map = new int[N];
        for (int i=0; i<N; i++){
            map[i] = Integer.parseInt(br.readLine());
        }
    }
}
